rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+ │tsh│                     │Load Balancer│                 │Teleport Proxy│
+ └─┬─┘                     └───────┬─────┘                 └───────┬──────┘
+   │                               │                               │
+   │ GET /webapi/connectionupgrade │                               │
+   │ Upgrade: alpn                 │                               │
+   ├──────────────────────────────►│                               │
+   │                               │  ┌───────────────┐            │
+   │                               ├──┤TLS Termination│            │
+   │                               │  └───────────────┘            │
+   │                               │                               │
+   │                               │ GET /webapi/connectionupgrade │
+   │                               │ Upgrade: alpn                 │
+   │                               ├──────────────────────────────►│
+   │                               │                               │
+   │                               │ HTTP 101 Switching Protocols  │
+   │                               │◄──────────────────────────────┤
+   │ HTTP 101 Switching Protocols  │                               │
+   │◄──────────────────────────────┤                               │
+   │                               │                               │
+   │ TLS Routing: teleport-mysql   │                               │
+   │◄──────────────────────────────┼──────────────────────────────►│
+   │                               │                               │
+```
+
+This allows the client to tunnel the "original TLS Routing" call through a Layer 7 (HTTP) load balancer.
+
+### Teleport Proxy with self-signed certs
+
+Load balancers usually terminate TLS with certificates signed by publicly trusted CAs. Since Teleport Proxy is no longer
+the public entry point, it is no longer neccessary for Teleport Proxy to serve a publicly trusted cert (e.g. using
+ACME). Thus Teleport Proxy should be welcome to use self-signed certs when sitting behind load balancers.
+
+To accommodate that, instead of serving the **Web** certs, the ALPN server on the Proxy will serve the **Host** certs
+when receiving connections from the upgrade flow. And instead of using the SystemCertPool, the Teleport client should
+use the **Host** CA for verifying the Proxy server when dialing TLS Routing inside an upgrade connection. Note that this
+modification does NOT apply to Teleport ALPN protocols that already do mTLS with other certs.
+
+### When to make a connection upgrade
+
+A new **optional** config parameter (e.g. `tsh_routing_upgrade_mode`) will be added in `proxy_service` section in the
+config yaml. This new config has no effect on how Proxy server serve incoming traffic but it will provide a hint to all
+the clients through the webapi ping:
+
+```
+{
+  "auth": {
+  ...
+  },
+  "proxy": {
+  ...
+    "tls_routing_enabled": true,
+    "tls_routing_upgrade_mode": "auto"
```

Comment:
Could we potentially combine these into a single parameter? For example:

```
tls_routing_mode: (on|off|alb)
```

Reply:
`alb` is quite narrow term. 
Also wonder if should detect if teleport proxy is behind L7 LB automaticity per tsh login call and make this option like "hard  enforcement" of TLS routing connection upgrade where we can still do the auto check  inside no latency crucial places like: `tsh login` call and allow to enforce Connection mode upgrade on all clients by this flag. 

This will allow a bit to simplify the UX teleport proxy configuration because it would be quite hard to describe all all possible situations when the `tls_routing_mode` should be set to `alb` mode.

Reply:
> tsh login call and allow to enforce Connection mode upgrade on all clients by this flag.
Ya, i was thinking about this too. we can have the test result in the profile yaml per `tsh login`. Let me drop the mode then, and we can just cache the result where it's needed.

Reply:
I just had an idea for improving performance for the test by merging the `/webapi/ping` and ALPN test in one call.

When client calls `/webapi/ping`, it will put these in ALPN `[]string{"teleport-webapi-ping", "http/1.1"}`.

The Proxy server will serve both ALPN values with the same response. 

The only difference is the `NegotiatedProtocol` would be `teleport-webapi-ping` if there is no LB, and it would be `http/1.1` or empty if LB is terminating the TLS.

What do you guys think?

Reply:
removed the mode for simplicity and added a section for performance on the detection test.